File: src/tokamak/validator.py
"""QAQC validation agent.

A Validator grades a processed span against its original on a 0..1 scale.
Three axes are returned (logical fidelity, content fidelity, safety) plus a
combined `signal`. The signal is what lands in the dataset row's `signal`
column so downstream training can filter or weight by quality.

Mirrored validation runs N validators in parallel against the same input and
returns the *minimum* signal (worst-judge rule — one bad reading dominates,
matching the model card recommendation that high-risk data deserves strict
QA).
"""

from __future__ import annotations

import json
import logging
import re
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from typing import List, Optional

from . import prompts
from .llm_client import LLMClient
from .processor import SpanInput

logger = logging.getLogger("tokamak.validator")


@dataclass
class ValidationResult:
    signal: float                       # 0..1 — combined score (min of axes)
    logical_fidelity: float
    content_fidelity: float
    safety: float
    notes: str = ""
    raw: str = ""                       # raw model output, for audit
    per_judge: List[float] = field(default_factory=list)  # mirrored-judge scores


_JSON_OBJECT_RE = re.compile(r"\{.*\}", re.DOTALL)


def _parse_score(raw: str) -> ValidationResult:
    """Pull the first JSON object out of the model output and clamp scores."""
    match = _JSON_OBJECT_RE.search(raw)
    if not match:
        logger.warning("validator: no JSON object found in output; defaulting to 0.0")
        return ValidationResult(0.0, 0.0, 0.0, 0.0, notes="parse_error", raw=raw)
    try:
        obj = json.loads(match.group(0))
    except json.JSONDecodeError as exc:
        logger.warning("validator: JSON decode failed (%s); defaulting to 0.0", exc)
        return ValidationResult(0.0, 0.0, 0.0, 0.0, notes="parse_error", raw=raw)

    def _f(key: str) -> float:
        try:
            v = float(obj.get(key, 0.0))
        except (TypeError, ValueError):
            return 0.0
        if v != v:  # NaN
            return 0.0
        return max(0.0, min(1.0, v))

    logical = _f("logical_fidelity")
    content = _f("content_fidelity")
    safety = _f("safety")
    # If the model returned its own signal, accept it; else use worst-axis.
    if "signal" in obj:
        signal = _f("signal")
    else:
        signal = min(logical, content, safety)
    notes = str(obj.get("notes", "")).strip()
    return ValidationResult(
        signal=signal,
        logical_fidelity=logical,
        content_fidelity=content,
        safety=safety,
        notes=notes,
        raw=raw,
    )


def _format_validate_user(span: SpanInput, processed: str, mode: str) -> str:
    return (
        "PROBLEM:\n"
        f"{span.problem.strip() or '(none)'}\n\n"
        "ANSWER:\n"
        f"{span.answer.strip() or '(none)'}\n\n"
        f"MODE: {mode}\n\n"
        "ORIGINAL:\n"
        f"{span.reasoning}\n\n"
        "PROCESSED:\n"
        f"{processed}"
    )


# ---------------------------------------------------------------------------
# Validator
# ---------------------------------------------------------------------------

class Validator:
    """Grade a processed span on 0..1. Optionally mirror N judges and take the
    worst score for robustness."""

    def __init__(
        self,
        *,
        model: Optional[str] = None,
        client: Optional[LLMClient] = None,
        mirrors: int = 1,
    ) -> None:
        if mirrors < 1:
            raise ValueError("mirrors must be >= 1")
        self.mirrors = mirrors
        # Slightly tighter cap on validation max_tokens — JSON output is tiny.
        self._client = client if client is not None else LLMClient(
            model=model, max_tokens=512
        )
        self.system = prompts.validate_prompt()

    def validate(self, span: SpanInput, processed: str, mode: str) -> ValidationResult:
        if self.mirrors == 1:
            return self._single(span, processed, mode)
        with ThreadPoolExecutor(max_workers=self.mirrors) as ex:
            results = list(
                ex.map(lambda _i: self._single(span, processed, mode),
                       range(self.mirrors))
            )
        # Mirrored: worst-judge rule for the headline signal.
        per_judge = [r.signal for r in results]
        worst_idx = min(range(len(results)), key=lambda i: results[i].signal)
        agg = results[worst_idx]
        agg.per_judge = per_judge
        return agg

    def _single(self, span: SpanInput, processed: str, mode: str) -> ValidationResult:
        user = _format_validate_user(span, processed, mode)
        try:
            raw = self._client.chat(system=self.system, user=user)
        except Exception as exc:  # noqa: BLE001
            logger.warning("validator call failed (%s); defaulting to 0.0", exc)
            return ValidationResult(0.0, 0.0, 0.0, 0.0, notes=f"call_error: {exc}")
        return _parse_score(raw)

File: rust/tokamak/src/validator.rs
//! QAQC validator agent. Three-axis grade (logical/content/safety) on 0..1,
//! signal = worst axis. Mirrored validators take worst-judge across N.

use std::sync::Arc;

use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::Value;
use tokio::sync::Semaphore;

use crate::llm::LlmClient;
use crate::processor::SpanInput;
use crate::prompts;

#[derive(Debug, Clone)]
pub struct ValidationResult {
    pub signal: f64,
    pub logical_fidelity: f64,
    pub content_fidelity: f64,
    pub safety: f64,
    pub notes: String,
    pub per_judge: Vec<f64>,
}

static JSON_OBJ_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(?s)\{.*\}").unwrap());

pub fn parse_score(raw: &str) -> ValidationResult {
    let Some(m) = JSON_OBJ_RE.find(raw) else {
        return ValidationResult {
            signal: 0.0, logical_fidelity: 0.0, content_fidelity: 0.0, safety: 0.0,
            notes: "parse_error".into(), per_judge: vec![],
        };
    };
    let Ok(v): Result<Value, _> = serde_json::from_str(m.as_str()) else {
        return ValidationResult {
            signal: 0.0, logical_fidelity: 0.0, content_fidelity: 0.0, safety: 0.0,
            notes: "parse_error".into(), per_judge: vec![],
        };
    };

    let clamp = |x: f64| if x.is_nan() { 0.0 } else { x.clamp(0.0, 1.0) };
    let f = |k: &str| -> f64 {
        v.get(k).and_then(|x| x.as_f64()).map(clamp).unwrap_or(0.0)
    };

    let logical = f("logical_fidelity");
    let content = f("content_fidelity");
    let safety  = f("safety");
    let signal = if v.get("signal").is_some() {
        f("signal")
    } else {
        logical.min(content).min(safety)
    };
    let notes = v.get("notes").and_then(|x| x.as_str()).unwrap_or("").to_string();

    ValidationResult { signal, logical_fidelity: logical, content_fidelity: content, safety, notes, per_judge: vec![] }
}

fn format_validate_user(s: &SpanInput, processed: &str, mode: &str) -> String {
    format!(
        "PROBLEM:\n{}\n\nANSWER:\n{}\n\nMODE: {}\n\nORIGINAL:\n{}\n\nPROCESSED:\n{}",
        if s.problem.trim().is_empty() { "(none)" } else { s.problem.trim() },
        if s.answer.trim().is_empty()  { "(none)" } else { s.answer.trim()  },
        mode, s.reasoning, processed,
    )
}

pub struct Validator { pub mirrors: usize }

impl Validator {
    pub fn new(mirrors: usize) -> Self { Self { mirrors: mirrors.max(1) } }

    /// Validate every (span, processed) pair concurrently. The total
    /// concurrency cap is `max_workers`; each item may further consume up
    /// to `mirrors` slots (one per mirrored judge).
    pub async fn validate_batch(
        &self,
        client: LlmClient,
        items: Vec<(SpanInput, String, &'static str)>,   // span, processed, mode_name
        max_workers: usize,
    ) -> Vec<ValidationResult> {
        if items.is_empty() { return Vec::new(); }
        let sem = Arc::new(Semaphore::new(max_workers.max(1)));
        let me = Arc::new(client);
        let mirrors = self.mirrors;

        let mut item_handles = Vec::with_capacity(items.len());
        for (span, processed, mode_name) in items.into_iter() {
            let sem = Arc::clone(&sem);
            let me  = Arc::clone(&me);
            item_handles.push(tokio::spawn(async move {
                grade_one(me, sem, mirrors, span, processed, mode_name).await
            }));
        }
        let mut out = Vec::with_capacity(item_handles.len());
        for h in item_handles {
            out.push(h.await.unwrap_or_else(|_| ValidationResult {
                signal: 0.0, logical_fidelity: 0.0, content_fidelity: 0.0, safety: 0.0,
                notes: "join_error".into(), per_judge: vec![],
            }));
        }
        out
    }
}

async fn grade_one(
    client: Arc<LlmClient>,
    sem: Arc<Semaphore>,
    mirrors: usize,
    span: SpanInput,
    processed: String,
    mode_name: &'static str,
) -> ValidationResult {
    let system = prompts::validate_prompt().to_string();
    let user = format_validate_user(&span, &processed, mode_name);
    let mut judge_handles = Vec::with_capacity(mirrors);
    for _ in 0..mirrors {
        let sem = Arc::clone(&sem);
        let client = Arc::clone(&client);
        let s = system.clone();
        let u = user.clone();
        judge_handles.push(tokio::spawn(async move {
            let _p = sem.acquire_owned().await.expect("semaphore");
            let raw = client.chat(&s, &u).await.unwrap_or_default();
            parse_score(&raw)
        }));
    }
    let mut judges: Vec<ValidationResult> = Vec::with_capacity(judge_handles.len());
    for h in judge_handles {
        if let Ok(r) = h.await { judges.push(r); }
    }
    if judges.is_empty() {
        return ValidationResult { signal: 0.0, logical_fidelity: 0.0, content_fidelity: 0.0, safety: 0.0,
                                  notes: "all_judges_failed".into(), per_judge: vec![] };
    }
    let per: Vec<f64> = judges.iter().map(|r| r.signal).collect();
    let worst = judges.into_iter().min_by(|a, b| a.signal.partial_cmp(&b.signal).unwrap()).unwrap();
    ValidationResult { per_judge: per, ..worst }
}
